contracts/sol6/reserves/KyberFprReserveV2.sol

```diff
@@ -0,0 +1,389 @@
+    using SafeMath for uint256;
+
+    mapping(bytes32 => bool) public approvedWithdrawAddresses; // sha3(token,address)=>bool
+    mapping(address => address) public tokenWallet;
+
+    struct ConfigData {
+        bool tradeEnabled;
+        bool doRateValidation; // whether to do rate validation in trade func
+        uint128 maxGasPriceWei;
+    }
+
+    address public kyberNetwork;
+    ConfigData internal configData;
+
+    IConversionRates public conversionRatesContract;
+    IKyberSanity public sanityRatesContract;
+    IWeth public weth;
+
+    event DepositToken(IERC20 indexed token, uint256 amount);
+    event TradeExecute(
+        address indexed origin,
+        IERC20 indexed src,
+        uint256 srcAmount,
+        IERC20 indexed destToken,
+        uint256 destAmount,
+        address payable destAddress
+    );
+    event TradeEnabled(bool enable);
+    event MaxGasPriceUpdated(uint128 newMaxGasPrice);
+    event DoRateValidationUpdated(bool doRateValidation);
+    event WithdrawAddressApproved(IERC20 indexed token, address indexed addr, bool approve);
+    event NewTokenWallet(IERC20 indexed token, address indexed wallet);
+    event WithdrawFunds(IERC20 indexed token, uint256 amount, address indexed destination);
+    event SetKyberNetworkAddress(address indexed network);
+    event SetConversionRateAddress(IConversionRates indexed rate);
+    event SetWethAddress(IWeth indexed weth);
+    event SetSanityRateAddress(IKyberSanity indexed sanity);
+
+    constructor(
+        address _kyberNetwork,
+        IConversionRates _ratesContract,
+        IWeth _weth,
+        uint128 _maxGasPriceWei,
+        bool _doRateValidation,
+        address _admin
+    ) public Withdrawable3(_admin) {
+        require(_kyberNetwork != address(0), "kyberNetwork 0");
+        require(_ratesContract != IConversionRates(0), "ratesContract 0");
+        require(_weth != IWeth(0), "weth 0");
+        kyberNetwork = _kyberNetwork;
+        conversionRatesContract = _ratesContract;
+        weth = _weth;
+        configData = ConfigData({
+            tradeEnabled: true,
+            maxGasPriceWei: _maxGasPriceWei,
+            doRateValidation: _doRateValidation
+        });
+    }
+
+    receive() external payable {
+        emit DepositToken(ETH_TOKEN_ADDRESS, msg.value);
+    }
+
+    function trade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) external override payable returns (bool) {
+        require(msg.sender == kyberNetwork, "wrong sender");
+        require(configData.tradeEnabled, "trade not enable");
+        require(tx.gasprice <= uint256(configData.maxGasPriceWei), "gas price too high");
+
+        doTrade(
+            srcToken,
+            srcAmount,
+            destToken,
+            destAddress,
+            conversionRate,
+            configData.doRateValidation
+        );
```

Comment:
tested with remix
See this file.
https://gist.github.com/ilanDoron/5ffd42ff4424d17e2eb41e3df1344e16

seems trade2 is best for gas.

Reply:
hm I see, the second was even higher in the js test I wrote, but it seems to be lower when testing with real deployed contract. Let me change to second method

Reply:
maybe the 3rd parameter in the struct made some difference